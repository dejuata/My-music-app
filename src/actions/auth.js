import {
    getAuth,
    signInWithPopup,
    createUserWithEmailAndPassword,
    updateProfile,
    signInWithEmailAndPassword,
    signOut
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { types } from "../types/types";
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2'


export const startLoginEmailPassword = (email, password) =>{

    return (dispatch) =>{

        dispatch( startLoading() );

        const auth = getAuth();
        signInWithEmailAndPassword(auth, email, password)
            .then(({ user }) => {
                console.log('dispatch login')
                dispatch(login(user.uid, user.displayName, 'firebase', ''))
            })
            .catch( (err) => {
                console.log(err.code, err.message);
                dispatch( finishLoading() );
                Swal.fire('Error', err.message, 'error')
            })

    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        const auth = getAuth();

        createUserWithEmailAndPassword(auth, email, password)
            .then( async ({ user }) => {

                await updateProfile(auth.currentUser, {
                    displayName: name,  photoURL: ""
                })

                dispatch(login(user.uid, user.displayName, 'firebase', ''))

            })
            .catch( (err) => {
                console.log(err.code, err.message);
                Swal.fire('Error', err.message, 'error')
            })

        console.log(auth.currentUser)
    }
}

export const startGoogleLogin = () =>{
    return (dispatch) =>{
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then( res => {
            })
    }
}

export const login = (uid, name, type, token) => ({
    type: types.login,
    payload: {
        uid,
        name,
        type,
        token
    }
})

export const startLogout = () => {
    return async (dispatch) => {
        const auth = getAuth();
        await signOut(auth)
        dispatch( logout() );
    }
}

export const logout = () => ({
    type: types.logout
})
