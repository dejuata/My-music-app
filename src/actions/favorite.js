import { types } from "../types/types";

export const setFavorites = (payload) => ({
    type: types.setFavorites,
    payload
})

export const removeFavorite= (payload) => ({
    type: types.removeFavorite,
    payload
})