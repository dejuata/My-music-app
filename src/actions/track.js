import { types } from "../types/types";

export const setTracks = (tracks) => ({
    type: types.setTracks,
    payload: tracks
})

export const addTrackFavorite = (id) => ({
    type: types.addTrackFavorite,
    payload: id
})

export const removeTrackFavorite = (id) => ({
    type: types.removeTrackFavorite,
    payload: id
})

export const logoutTrack = () => ({
    type: types.logoutTrack
})