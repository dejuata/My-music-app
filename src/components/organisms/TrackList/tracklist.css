@media (min-width: 768px) {
    .card-container {
        align-items: center;
        column-gap: 20px;
        display: flex;
        flex-wrap: wrap;
        flex: 0 1 calc(33% 1em);
        justify-content: center;
        row-gap: 20px;
        width: calc(100% + 16px);
    }
}