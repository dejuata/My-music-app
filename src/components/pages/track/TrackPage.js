import React from 'react'
import { useParams } from 'react-router-dom';
import { useTrackDetail } from '../../../hooks/useTrackDetail';
import { TrackDetail } from '../../organisms/TrackDetail/TrackDetail';


export const TrackPage = () => {

    const { trackId } = useParams();

    const { data:track, loading } = useTrackDetail(trackId);


    return (
        <>
            {

                loading
                ?
                    <p>Loading</p>
                :
                    <TrackDetail {...track}/>

            }
        </>
    )
}
