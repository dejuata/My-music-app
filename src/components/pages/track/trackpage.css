.track-detail {
    width: 100%;
    /* background-color: var(--color-white); */
    /* color: var(--color-dark); */
    border-radius: 12px;
    padding: 12px;
    box-sizing: border-box;
}

.track-detail__image {
    margin: 0;
    /* padding: 20px; */
    /* box-sizing: border-box; */
}

.track-detail__image img{
    width: 100%;
    object-fit: cover;
    border-radius: 12px;
}

.track-detail__info {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    justify-content: space-between;
    margin-top: 24px;
}

.track-detail__info h2 {
    margin: 0 ;
    font-size: var(--xlg);
    font-weight: bold;
}

.track-detail__info p {
    margin: 0;
}

.track-detail__info p:nth-child(2) {
    margin-bottom: 32px;
    font-size: var(--lg);

}

.track-detail__info p:nth-child(3) {
    font-size: var(--md);
    text-align: justify;
    font-weight: 300;
    margin-bottom: 12px;

}

.track-detail__info p:nth-child(4) {
    margin-bottom: 32px;
    font-weight: 300;
    font-size: var(--sm);
}

.secondary-button {
    background-color: var(--color-white);
    border-radius: 8px;
    border: 1px solid var(--color-text);
    color: var(--color-green);
    cursor: pointer;
    font-size: var(--lg);
    font-weight: 500;
    text-align: center;
    padding: 20px;
    box-sizing: border-box;
    width: 100%;
    text-decoration: none;
}