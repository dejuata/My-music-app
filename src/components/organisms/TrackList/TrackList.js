import React from 'react'
import { useDispatch, useSelector } from 'react-redux';
import { removeFavorite } from '../../../actions/favorite';
import { addTrackFavorite, removeTrackFavorite } from '../../../actions/track';
import { deleteFavoriteTrack } from '../../../api/selectors/deleteFavoriteTrack';
import { putFavoriteTrack } from '../../../api/selectors/putFavoriteTrack';
import { TrackCard } from '../../molecules/TrackCard/TrackCard'


export const TrackList = ({ tracks }) => {

    const dispatch = useDispatch();

    const handleActive = (id) => {
        putFavoriteTrack(id)
        dispatch( addTrackFavorite(id) );
    }

    const handleDeactive = (id) => {
        deleteFavoriteTrack(id)
        dispatch( removeFavorite(id) );
        dispatch( removeTrackFavorite(id) );
    }

    return (
        <div className='o-track-list-container'>
            {
                tracks.map( track => {
                    return <TrackCard
                                key={ track.id }
                                track={ track }
                                handleActive={ () => handleActive(track.id) }
                                handleDeactive={ () => handleDeactive(track.id) }
                            />
                })
            }
        </div>
    )
}
