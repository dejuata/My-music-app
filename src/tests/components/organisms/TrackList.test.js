import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';

import { demoTracks } from '../../fixtures/demoTracks';
import { TrackList } from '../../../components/organisms/TrackList/TrackList';
import { Provider } from 'react-redux';
import { addTrackFavorite, removeTrackFavorite } from '../../../actions/track';
import { demoFavorites } from '../../fixtures/demoFavorites';
import { removeFavorite } from '../../../actions/favorite';

jest.mock('../../../actions/track', () => ({
    addTrackFavorite: jest.fn(),
    removeTrackFavorite: jest.fn()
}))

jest.mock('../../../actions/favorite', () => ({
    removeFavorite: jest.fn(),
}))


const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);
store.dispatch = jest.fn()

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <TrackList
                tracks={ demoTracks.tracks }
            />
        </MemoryRouter>
    </Provider>
)

describe('Test TrackList', () => {

    beforeEach( () => {
        store = mockStore(initState);
        jest.clearAllMocks();
    })

    test('Should be render TrackList', () => {

        expect(wrapper).toMatchSnapshot();

    })

    test('Should be render two <TrackCard />', () => {

        expect( wrapper.find('TrackCard').length ).toBe( demoTracks.tracks.length )

    })

    test('Should be run the action addTrackFavorite', () => {

        let store = mockStore(initState);
        store.dispatch = jest.fn()

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <TrackList
                        tracks={ [ demoTracks.tracks[0] ] }
                    />
                </MemoryRouter>
            </Provider>
        )

        wrapper.find('FavoriteButton').simulate('click')

        expect( addTrackFavorite ).toHaveBeenCalledWith('1');

    })

    test('Should be run the action removeTrackFavorite', () => {

        let store = mockStore(initState);
        store.dispatch = jest.fn()
        let track = demoTracks.tracks[0];
        track.favorite = true;

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <TrackList
                        tracks={ [ track ] }
                    />
                </MemoryRouter>
            </Provider>
        )

        wrapper.find('FavoriteButton').simulate('click')

        expect( removeTrackFavorite ).toHaveBeenCalledWith('1');

    })

    test('Should be run the action removeFavorite', () => {

        let store = mockStore(initState);
        store.dispatch = jest.fn()

        const wrapper = mount(
            <Provider store={ store }>
                <MemoryRouter>
                    <TrackList
                        tracks={ [ demoFavorites.tracks[0] ] }
                    />
                </MemoryRouter>
            </Provider>
        )

        wrapper.find('FavoriteButton').simulate('click')

        expect( removeFavorite ).toHaveBeenCalledWith('1');

    })

})