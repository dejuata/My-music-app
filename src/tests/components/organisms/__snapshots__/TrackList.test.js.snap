// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test TrackList Should be render TrackList 1`] = `
<div
  className="o-track-list-container"
>
  <div
    className="m-card-track animate__animated animate__fadeIn "
  >
    <div
      className="m-card-track__body"
    >
      <a
        href="/track/1"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <figure
          className="m-card-track__body-image"
        >
          <img
            alt="Voy Pa'Encima"
            src="https://i.scdn.co/image/ab67616d00001e021386b29ab009dd27c398d141"
          />
        </figure>
        <div
          className="m-card-track__body-detail"
        >
          <p>
            Desn√∫date Mujer
          </p>
          <p>
            Frankie Ruiz
          </p>
        </div>
      </a>
    </div>
    <div
      className="m-card-track__footer"
    >
      <span>
        4:47
      </span>
      <button
        className="a-favorite-button"
        onClick={[Function]}
      >
        <svg
          className=" a-favorite__icon false"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            key="0"
          />
        </svg>
      </button>
    </div>
  </div>
  <div
    className="m-card-track animate__animated animate__fadeIn "
  >
    <div
      className="m-card-track__body"
    >
      <a
        href="/track/2"
        onClick={[Function]}
        style={
          Object {
            "textDecoration": "none",
          }
        }
      >
        <figure
          className="m-card-track__body-image"
        >
          <img
            alt="Vendedora de Caricias"
            src="https://i.scdn.co/image/ab67616d00001e02ef4bb48191b9d3f5fd779324"
          />
        </figure>
        <div
          className="m-card-track__body-detail"
        >
          <p>
            Vendedora de Caricias
          </p>
          <p>
            Panteon Rococo
          </p>
        </div>
      </a>
    </div>
    <div
      className="m-card-track__footer"
    >
      <span>
        3:45
      </span>
      <button
        className="a-favorite-button"
        onClick={[Function]}
      >
        <svg
          className=" a-favorite__icon false"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
            key="0"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
